### PR TITLE
bugfix- expose the legend and the hint to the template

### PR DESCRIPTION
**Pull request by sulthan-ahmed**

### CODE REVIEW

**Comment by sulthan-ahmed** [comment]
Is there a need to expose a label?

**Comment by joefitter** [comment]
no - a label isn't a "thing" - it uses a legend as it is in a `fieldset` describing a group of fields

File: lib/date.js [hunk 1]
```diff
@@ -98,8 +98,18 @@ module.exports = (key, options) => {
...
+    const legendKey = `fields.${key}.legend`;
+    let legend = req.translate(legendKey);
+    if (legend === legendKey) {
+      legend = null;
+    }
+    const hintKey = `fields.${key}.hint`;
+    let hint = req.translate(hintKey);
+    if (hint === hintKey) {
+      hint = null;
+    }
```
Comment (joefitter) [inline lib/date.js]: Just to save on repetition, you could extract the conditional translation logic to a helper function:

```js
const conditionalTranslate = (key, translate) => {
  let result = translate(key);
  if (result === key) {
    result = null;
  }
  return result;
};

// then use it like:
const legend = conditionalTranslate(`fields.${key}.legend`, req.translate);
const hint = conditionalTranslate(`fields.${key}.hint`, req.translate);
```

**Review by joefitter — CHANGES_REQUESTED** [review]

File: test/date.js [hunk 2]
```diff
@@ -153,7 +153,15 @@ describe('Date Component', () => {
...
-        expect(res.render).to.have.been.calledWith(path.resolve(__dirname, '../templates/date.html'), { error, key: 'date-field' });
+        expect(res.render).to.have.been.calledWith(path.resolve(__dirname, '../templates/date.html'), { error, key: 'date-field', legend: null, hint: null });
+      });
+
+      it('calls res.render with template, key, legend and hint with values NOT null', () => {
+        req.translate = sinon.stub();
+      req.translate.withArgs('fields.date-field.legend').returns('you legend');
+      req.translate.withArgs('fields.date-field.hint').returns('some hint');
+      date.hooks['pre-render'](req, res, next);
+      expect(res.render).to.have.been.calledWith(path.resolve(__dirname, '../templates/date.html'), { error: undefined, key: 'date-field', legend: 'you legend', hint: 'some hint' });
```
Comment (lennym) [inline test/date.js]: You've ended up with wacky indentation again.

**Review by lennym — CHANGES_REQUESTED** [review]

File: test/date.js (hunk 2 above)
Comment (sulthan-ahmed) [reply test/date.js]: woops, this time I can't blame sublime.  It was when I rebased

File: lib/date.js (hunk 1 above)
Comment (sulthan-ahmed) [reply lib/date.js]: looks awesome thanks man

File: lib/date.js [hunk 3]
```diff
@@ -98,8 +98,17 @@ module.exports = (key, options) => {
         label: label === labelKey ? v.label : label
       });
     }));
+    const conditionalTranslate = (fieldKey) => {
+      let result = req.translate(fieldKey);
+      if (result === fieldKey) {
+        result = null;
+      }
+      return result;
+    };
```
Comment (joefitter) [inline lib/date.js]: only a _minor_ thing - but I would move this function outside of the main export function - up where you see the other helpers. This is because here it has access to all of the internal variables such as key, fields etc... This could lead to some hard to find bugs if you rely on an outside variable etc. I guess this is why you renamed `key` to `fieldKey` (which doesn't really make sense) - if you move it outside of the module export function scope, you can go back to calling it `key` - which I prefer :)

**Review by joefitter — CHANGES_REQUESTED** [review]

File: lib/date.js (hunk 3 above)
Comment (joefitter) [reply lib/date.js]: this is a `pure` function - meaning whenever it is called with the same arguments, it will return the same value. This is the case here, but if you were to rely on an external argument not passed into the function, it would no longer be pure - an example of this would be:

```js
const key = 'field-name';

const conditionalRender = (part, translate) => {
  const translationKey = `fields.${key}.${part}`;
  let result = translate(translationKey);
  if (result === translationKey) {
    result = null;
  }
  return result;
}
```

^ This function relies on the external constant `key`, the function will return a different value if called with the same arguments but `key` is changed.

To sum up, impure functions are more bug-prone as they are dependent on external variables/factors

Comment (sulthan-ahmed) [reply lib/date.js]: Totally makes sense about having it outside of the scope.  Now I see why you have two arguments including the translate function.

Regarding pure functions just been reading about it.  So is it basically that the function can't rely on other external functions or variables?
I also read it shouldn't have side effects and mutate shared things

Comment (joefitter) [reply lib/date.js]: exactly that! Pure functions enforce much simpler code, as you can see exactly what the function has access to. Helpers like this perform one action and don't have any side effects.

```js
getValues(req, res, next) {
  const getVal = () => {
    return req.sessionModel.get(req.params.value);
  }
  getVal()
}

// vs

const getVal = (val, model) => {
  return model.get(val);
}

class blah {
  getValues(req, res, next) {
    const val = getVal(req.params.value, req.sessionModel);
  }
}
```

**Review by joefitter — APPROVED** [review]

**Comment by sulthan-ahmed** [comment]
🎉  thanks dude
